File: src/main/java/org/scalatest/spi/location/FreeSpecFinder.java
package org.scalatest.spi.location;

import java.util.ArrayList;
import java.util.List;

public class FreeSpecFinder implements Finder {

    private String getTestNameBottomUp(MethodInvocation invocation) {
        AstNode parent = invocation.getParent();
        AstNode target = invocation.getTarget();
        if (parent == null || !(parent instanceof MethodInvocation)) 
            return target.toString();
        else 
            return getTestNameBottomUp((MethodInvocation) parent) + " " + target.toString();
    }
    
    private List<String> getTestNamesUpBottom(MethodInvocation invocation) {
        String prefix = getTestNameBottomUp(invocation);
        AstNode[] children = invocation.getChildren();
        List<String> testNameList = new ArrayList<String>();
        for(AstNode childNode : children) {
            if(childNode instanceof MethodInvocation) {
                MethodInvocation child = (MethodInvocation) childNode;
                if (child.getName().equals("in") || child.getName().equals("is")) 
                    testNameList.add(prefix + " " + child.getTarget().toString());
                else if (child.getName().equals("-"))
                    testNameList.addAll(getTestNamesUpBottom(child));
            }
        }
        return testNameList;
    }
    
    public Test find(AstNode node) {
        Test test = null;
        if (node instanceof MethodInvocation) {
            MethodInvocation invocation = (MethodInvocation) node;
            if (invocation.getName().equals("in") || invocation.getName().equals("is")) {
                String testName = getTestNameBottomUp(invocation);
                test = new Test(invocation.getClassName(), testName, new String[] { testName });
            }
            else if (invocation.getName().equals("-")) {
                String displayName = getTestNameBottomUp(invocation);
                List<String> testNameSet = getTestNamesUpBottom(invocation);
                test = new Test(invocation.getClassName(), displayName, testNameSet.toArray(new String[testNameSet.size()]));
            }
        }    
        return test;
    }
    
}

File: src/main/scala/org/scalatest/finders/FreeSpecFinder.scala
package org.scalatest.finders

import scala.collection.mutable.ListBuffer
import scala.annotation.tailrec

class FreeSpecFinder extends Finder {
  
  private def getTestNameBottomUp(invocation: MethodInvocation): String = {
    if(invocation.parent == null || !invocation.parent.isInstanceOf[MethodInvocation])
      invocation.target.toString
    else
      getTestNameBottomUp(invocation.parent.asInstanceOf[MethodInvocation]) + " " + invocation.target.toString  
  }
  
  private def getTestNamesTopDown(invocation: MethodInvocation): List[String] = {
    @tailrec
    def getTestNamesTopDownAcc(invocations: List[AstNode], acc: List[String]): List[String] = invocations match {
      case Nil => acc
      case node :: rs => 
        node match {
          case invocation: MethodInvocation => 
            if (invocation.name == "in" || invocation.name == "is") {
              val testName = getTestNameBottomUp(invocation)
              getTestNamesTopDownAcc(rs, testName :: acc)
            }
            else {
              getTestNamesTopDownAcc(invocation.children.toList ::: rs, acc)
            }
          case _ => getTestNamesTopDownAcc(rs, acc)
        }
    }
    getTestNamesTopDownAcc(List(invocation), List.empty).reverse
  }
  
  def find(node: AstNode): Option[Selection] = {
    node match {
      case invocation @ MethodInvocation(className, target, parent, children, name, args) =>
        name match {
          case "in" | "is" if parent.name == "-" =>
            val testName = getTestNameBottomUp(invocation)
            Some(new Selection(className, testName, Array(testName)))
          case "-" =>
            val displayName = getTestNameBottomUp(invocation)
            val testNames = getTestNamesTopDown(invocation)
            Some(new Selection(className, displayName, testNames.toArray))
          case _ => 
            if (node.parent != null)
              find(node.parent)
            else
              None
        }
      case _ => 
        if (node.parent != null)
          find(node.parent)
        else
          None
    }
  }
}